File: templates/typescript/query-with-tools.ts
import { query } from "@anthropic-ai/claude-agent-sdk";

/**
 * Query with Tool Control Template
 *
 * Demonstrates:
 * - allowedTools / disallowedTools
 * - System prompts
 * - Handling assistant messages with API message objects
 */

async function queryWithTools() {
  for await (const message of query({
    prompt: "Review the auth module for security issues and fix vulnerabilities",
    options: {
      workingDirectory: "/path/to/project",
      systemPrompt: `You are a security-focused code reviewer.
Analyze code for SQL injection, XSS, auth bypass, and OWASP issues.`,

      allowedTools: ["Read", "Grep", "Glob", "Write", "Edit"],
      disallowedTools: ["Bash"],
    },
  })) {
    switch (message.type) {
      case "assistant":
        // message.message is the full Anthropic API APIAssistantMessage
        for (const block of message.message.content) {
          if (block.type === "text") {
            console.log(block.text);
          } else if (block.type === "tool_use") {
            console.log(`Tool: ${block.name}`);
          }
        }
        break;

      case "result":
        if (message.subtype === "success") {
          console.log(`\nResult: ${message.result}`);
          console.log(`Cost: $${message.total_cost_usd.toFixed(4)}`);
        } else {
          console.error(`Error: ${message.subtype}`, message.errors);
        }
        break;
    }
  }
}

queryWithTools().catch(console.error);

File: templates/python/query_with_tools.py
"""Claude Agent SDK query with custom tools."""
import asyncio
from typing import Any
from claude_agent_sdk import (
    query, tool, create_sdk_mcp_server, ClaudeAgentOptions,
    ResultMessage,
)

@tool("get_weather", "Get weather for a city", {"city": {"type": "string", "description": "City name"}})
async def get_weather(args: dict[str, Any]) -> dict[str, Any]:
    city = args.get("city", "")
    return {"content": [{"type": "text", "text": f"Weather in {city}: 72°F, sunny"}]}

async def main():
    server = create_sdk_mcp_server(name="weather", tools=[get_weather])

    options = ClaudeAgentOptions(
        mcp_servers={"weather": server},
        permission_mode="bypassPermissions",
        max_turns=10,
    )

    async for message in query(prompt="What's the weather in Tokyo?", options=options):
        if isinstance(message, ResultMessage) and message.subtype == "success":
            print(message.result)

asyncio.run(main())
